server/src/main/java/org/elasticsearch/search/aggregations/InternalAggregation.java

```diff
@@ -125,9 +128,15 @@ public PipelineTree pipelineTreeRoot() {
          * the maximum number of buckets allowed in a response
          */
         public void consumeBucketsAndMaybeBreak(int size) {
+            if (isCanceled.get()) {
```

Comment:
It's worth adding a comment that this is a volatile read. Do we want to do that on every consume? We don't check the memory breaker on every consume so maybe we should do this either?

I see that `AtomicBoolean` has a bunch of fun new ways to fetch the value without memory effects - but only in java 9, so, sadly, probably not for us yet.

Reply:
My reasoning here was that this happens in the reduce context and as you said, it's AtomicBoolean.get(). Do you think it will have noticeable overhead here? I can probably move inside MultiBucketConsumer and call it every few buckets if you think it makes sense.

Reply:
I'm not really sure! Its the kind of location that could suffer from it but I don't know how much without benchmarks. I'd feel quite comfortable without benchmarks if it were in the same 1024 calls thing, but it wouldn't interrupt as immediately. I expect the checks are much lighter than the circuit breaker checks but heavier than the general bucket building stuff.

Reply:
Thanks! I will try to run some benchmarks to see what's going on there.